source_pubsub/client.go

```diff
@@ -255,39 +259,56 @@ func (c *PubSubClient) Wait(ctx context.Context) error {
 	}
 }
 
+func (c *PubSubClient) ResetRetryCount() {
+	c.retryCount = c.maxRetries
+}
+
 func (c *PubSubClient) ReplayID() []byte {
 	return c.currReplayID
 }
 
+func (c *PubSubClient) retryAuth(ctx context.Context, retry bool) (error, bool) {
+	for retry {
+		var err error
+		sdk.Logger(ctx).Info().Msgf("retry connection - retries %d ", c.retryCount)
+
+		if err = c.login(ctx); err != nil && c.retryCount <= 0 {
+			return fmt.Errorf("failed to refresh auth: %w", err), retry
+		} else if err != nil {
+			sdk.Logger(ctx).Info().Msgf("received error on login - retry - %d ", c.retryCount)
+			c.retryCount--
+		}
+
+		if err = c.Initialize(ctx); err != nil && c.retryCount <= 0 {
```

Comment:
This will not work, because Initialize can only be called once, it is not setup to be re-entrant.

Firstly it will start another go-routine to subscribe (starting startCDC from within startCDC)

https://github.com/conduitio-labs/conduit-connector-salesforce/blob/a30d12b09cc2fa9d34d3485300975c881a5062ff/source_pubsub/client.go#L146-L152

Then it launches another go-routine to wait for the tomb to die

https://github.com/conduitio-labs/conduit-connector-salesforce/blob/a30d12b09cc2fa9d34d3485300975c881a5062ff/source_pubsub/client.go#L154-L159

It will be hairy to introduce a full client retry within the client itself.

My suggestion is to handle retry-auth errors within the code since these do not affect the gRPC stream but rather just the headers. The auth is simply involved in getting new tokens which are then passed to each call. 

If the gRPC connection is at this point broken, this will need to be reinitialized:

https://github.com/conduitio-labs/conduit-connector-salesforce/blob/a30d12b09cc2fa9d34d3485300975c881a5062ff/source_pubsub/client.go#L101-L104

With these two variables reset https://github.com/conduitio-labs/conduit-connector-salesforce/blob/a30d12b09cc2fa9d34d3485300975c881a5062ff/source_pubsub/client.go#L120-L121

An alternative is to keep original retry auth logic, but create another method `reconnect(...)` on `PubSubClient` which will close the gRPC connection, reconnect and re-assign the new objects.

Reply:
We've been calling Initizalize from startCDC before already - https://github.com/conduitio-labs/conduit-connector-salesforce/blob/10e02844f1ae6b61e5fcea7947c5d4fe5f40ccb1/source_pubsub/client.go#L286 
Why would this cause issues over what he have?

Reply:
yes that is wrong @nassor we need to back out that change.

because it will recursively start go-routines inside the tomb. (x(x(x....)))

Reply:
It is probably not used as a `tomb` at all (why do we use that in the first place?).